### PR TITLE
Revised approach to ATM Project. Details in Desc

### PR DESCRIPTION
**Pull request by rehanbrr**
ATM project using MVC.

- User model has name, email, password and username that can be saved
- User controller manages the user model. Actions like creating user, logging in, changing password and getting username or email for validation are done here.

- Account model uses user's username to identify which user it belongs to
- Account controller handles all the actions for account which include retrieving all the accounts that user owns. Features like depositing money, withdrawing and deleting account are yet to be added

- FileHandling file is responsible for retrieving data when application starts from the csv files of user and account. It is also responsible for saving, for now. Saving will be changed to only happen when the application is exited, saving only the newer entries.

- Views is the frontend of the application where all display calls are made. The views file interacts with the controllers to get information to display.

- Users and Accounts of each user are stores in a hash -- account hash has the user's email as key and array of user's accounts as value

**Update**

Implemented withdraw, deposit, check balance and delete account features to each account
Working on optimising code, ability to send money to other users and changing user password

**Update Final**

Optimised code as much as I could. Implemented sending money and changing password. Need review

### CODE REVIEW

File: AccountController.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  def create_account(pin, username)
+    account = Account.new(pin, username)
+    @account_data << account
+  end
+
+  def find_user_accounts(username)
+    accounts = @account_data.select {|account| account.username == username}
+
+    return accounts
```
Comment (aminasaif0) [inline AccountController.rb]: no need for return statement

File: FileHandling.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+class FileSystem
+
+  attr_accessor :user_data, :account_data
+
+  def load_user_data
+    file_name = 'users.csv'
+    if File.exist?(file_name)
+      user_file = File.open(file_name, 'r')
+      user_data = user_file.readlines
+      user_file.close
```
Comment (aminasaif0) [inline FileHandling.rb]: we can read and seperate data in the same block

File: FileHandling.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      @user_data = []
+
+      for user_line in user_data
+        user_params = user_line.split(',')
+        user = User.new(user_params[0], user_params[1], user_params[2], user_params[3])
+        @user_data << user
+      end
+    else
+      File.new(file_name, 'w+')
+      return nil
```
Comment (aminasaif0) [inline FileHandling.rb]: no need to explicitly write return

File: FileHandling.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      user_file.close()
+    end
+
+  end
+
+  def load_account_data
+    file_name = 'accounts.csv'
+
+    if File.exist?(file_name)
+      account_file = File.open(file_name, 'r')
```
Comment (aminasaif0) [inline FileHandling.rb]: read and seperate assign data in the same go

File: FileHandling.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      @account_data = []
+
+      for account_line in account_data
+        account_params = account_line.split(',')
+        account = Account.new(account_params[0], account_params[1], account[2])
+        @account_data << account
+      end
+    else
+      File.new(file_name, 'w+')
+      return nil
```
Comment (aminasaif0) [inline FileHandling.rb]: no need to explicitly write return

**Review by aminasaif0 — CHANGES_REQUESTED** [review]

File: UserController.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  def start_user_controller
+    @file_system = FileSystem.new
+    @file_system.load_user_data
+    @user_data = @file_system.user_data
+  end
+
+  def get_emails
+    @user_data.map do |user|
+      user.email
```
Comment (aminasaif0) [inline UserController.rb]: use pluck if you are retrieving one column from active record instead of map.

File: UserController.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  def get_usernames
+    @user_data.map do |user|
+        user.username
```
Comment (aminasaif0) [inline UserController.rb]: use pluck if you are retrieving one column from active record instead of map.

File: UserController.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  def change_password(password)
+    if @user
+      @user.password = password
+    end
```
Comment (aminasaif0) [inline UserController.rb]: @user&.password = password

File: UserController.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def create_user(name, email, password, username)
+    user = User.new(name, email, password, username)
+    @user_data << user
```
Comment (aminasaif0) [inline UserController.rb]: why assign user to instance variable in create ?

**Review by aminasaif0 — CHANGES_REQUESTED** [review]

File: UserController.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def login_user(email, password)
+    puts "this is userdata #{@user_data}"
+    @user = @user_data.select {|user| user.email == email && user.password == password}
```
Comment (aminasaif0) [inline UserController.rb]: why return the array instead of one user which matches the input data?

File: UserController.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    puts "this is #{@user}"
+    return true unless @user.nil?
```
Comment (aminasaif0) [inline UserController.rb]: return !@user.nil?

File: validation.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
+module Validation
+  def email_valid?(email)
+    emails = get_emails
```
Comment (aminasaif0) [inline validation.rb]: why do we have this before we check the validity of the email?

File: validation.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    if email.include?('@') and email.include?('.com') and !email.include?(',')
+      return emails.any? {|taken_email| taken_email == email}
+    end
+    
+    return true
+  end
+  
+  def password_valid?(password)
+    if password.length > 8 and password =~ /\d/ and !password.include?(',')
```
Comment (aminasaif0) [inline validation.rb]: refactor

**Review by aminasaif0 — CHANGES_REQUESTED** [review]

File: AccountController.rb (hunk 1 above)
Comment (rehanbrr) [reply AccountController.rb]: removed

File: FileHandling.rb (hunk 2 above)
Comment (rehanbrr) [reply FileHandling.rb]: done

File: FileHandling.rb (hunk 3 above)
Comment (rehanbrr) [reply FileHandling.rb]: removed

File: FileHandling.rb (hunk 4 above)
Comment (rehanbrr) [reply FileHandling.rb]: done

File: AccountController.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+  def add_account(account)
+    @account_data[account.email] ||= []
+    @account_data[account.email] << account
+  end
+
+  def generate_account_number
+    account_number = 0
+    loop do
+      account_number = rand(10000...99999).to_s
```
Comment (aminasaif0) [inline AccountController.rb:32]: why is account number random?

**Review by aminasaif0 — CHANGES_REQUESTED** [review]

File: AccountController.rb (hunk 14 above)
Comment (rehanbrr) [reply AccountController.rb:32]: I wanted the account numbers to be within a certain range and also banks don't give you an option to set your own account number, they just generate it randomly.

**Review by aminasaif0 — APPROVED** [review]

**Pull request merged**